### PR TITLE
[evm] add EIP-2930 access list transaction and enable EVM Berlin

### PR DESCRIPTION
**Pull request by dustinxie**
EIP-2930: https://eips.ethereum.org/EIPS/eip-2930 adds a new access list tx, the implementation was here: https://github.com/ethereum/go-ethereum/commit/bbfb1e4008a359a8b57ec654330c0e674623e52f

### CODE REVIEW

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 1]
```diff
@@ -541,6 +556,19 @@ func (stateDB *StateDBAdapter) RevertToSnapshot(snapshot int) {
...
+	{
+		delete(stateDB.accessListSnapshot, snapshot)
+		for i := snapshot + 1; ; i++ {
+			if _, ok := stateDB.accessListSnapshot[i]; ok {
+				delete(stateDB.accessListSnapshot, i)
+			} else {
+				break
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply action/protocol/execution/evm/evmstatedbadapter.go:571]: when revert to a snapshot, need to clear all snapshots later than it
same as line 550-557 right above

File: action/execution.go [hunk 2]
```diff
@@ -78,6 +82,44 @@ func (ex *Execution) Data() []byte { return ex.data }
 // Payload is same as Data()
 func (ex *Execution) Payload() []byte { return ex.data }
 
+// AccessList returns the access list
+func (ex *Execution) AccessList() types.AccessList { return ex.accessList }
+
+func toAccessListProto(list types.AccessList) []*iotextypes.AccessTuple {
+	if len(list) == 0 {
+		return nil
```
Comment (Liuhaai) [inline action/execution.go:90]: Is it ok to return `nil` instead of empty slice?

File: action/execution.go [hunk 3]
```diff
@@ -78,6 +82,44 @@ func (ex *Execution) Data() []byte { return ex.data }
...
+	}
+	proto := make([]*iotextypes.AccessTuple, len(list))
+	for i, v := range list {
+		proto[i] = &iotextypes.AccessTuple{}
+		proto[i].Address = hex.EncodeToString(v.Address.Bytes())
+		if numKey := len(v.StorageKeys); numKey > 0 {
+			proto[i].StorageKeys = make([]string, numKey)
```
Comment (Liuhaai) [inline action/execution.go:97]: same here

File: action/protocol/execution/evm/evm.go [hunk 4]
```diff
@@ -475,16 +492,20 @@ func evmErrToErrStatusCode(evmErr error, g genesis.Blockchain, height uint64) (e
...
+func intrinsicGas(size uint64, list types.AccessList) (uint64, error) {
 	if action.ExecutionDataGas == 0 {
 		panic("payload gas price cannot be zero")
 	}
-	dataSize := uint64(len(data))
-	if (math.MaxInt64-action.ExecutionBaseIntrinsicGas)/action.ExecutionDataGas < dataSize {
+
+	var accessListGas uint64
+	if len(list) > 0 {
+		accessListGas = uint64(len(list)) * action.TxAccessListAddressGas
```
Comment (Liuhaai) [inline action/protocol/execution/evm/evm.go:502]: `+=`

**Review by Liuhaai — APPROVED** [review]

File: action/execution.go (hunk 2 above)
Comment (dustinxie) [reply action/execution.go:90]: yeah, there is test case with `list = nil`

File: action/execution.go (hunk 3 above)
Comment (dustinxie) [reply action/execution.go:97]: yeah, there is test case with `v.StorageKeys = nil`

File: action/protocol/execution/evm/evm.go (hunk 4 above)
Comment (dustinxie) [reply action/protocol/execution/evm/evm.go:502]: here is the first assignment, so can use `=`, the next line use `+=`

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 5]
```diff
@@ -541,6 +556,19 @@ func (stateDB *StateDBAdapter) RevertToSnapshot(snapshot int) {
 			}
 		}
 	}
+	// restore access list
+	stateDB.accessList = nil
+	stateDB.accessList = stateDB.accessListSnapshot[snapshot]
+	{
```
Comment (CoderZhi) [inline action/protocol/execution/evm/evmstatedbadapter.go:562]: what's this bracket for?

Comment (dustinxie) [reply action/protocol/execution/evm/evmstatedbadapter.go:562]: better for visual clarity. other places are in a code block by `if stateDB.fixSnapshotOrder`, here no need for the `if`, but leave the bracket to show the same code block

**Pull request merged**